internal/controllers/machinedeployment/mdutil/util.go

```diff
@@ -469,58 +475,71 @@ func FindNewMachineSet(deployment *clusterv1.MachineDeployment, msList []*cluste
 	// We deterministically choose the oldest new MachineSet with matching template hash.
 	sort.Sort(MachineSetsByDecreasingReplicas(msList))
 
-	var matchingMachineSets []*clusterv1.MachineSet
+	var newMSCandidates []*clusterv1.MachineSet
+	oldMSs = make([]*clusterv1.MachineSet, 0)
+	oldMSNotUpToDateResults = make(map[string]NotUpToDateResult)
 	var diffs []string
 	for _, ms := range msList {
-		upToDate, logMessages, _ := MachineTemplateUpToDate(&ms.Spec.Template, &deployment.Spec.Template)
+		upToDate, notUpToDateResult := MachineTemplateUpToDate(&ms.Spec.Template, &deployment.Spec.Template)
 		if upToDate {
-			matchingMachineSets = append(matchingMachineSets, ms)
+			newMSCandidates = append(newMSCandidates, ms)
 		} else {
-			diffs = append(diffs, fmt.Sprintf("MachineSet %s: diff: %s", ms.Name, strings.Join(logMessages, ", ")))
+			oldMSs = append(oldMSs, ms)
+			oldMSNotUpToDateResults[ms.Name] = *notUpToDateResult
+			diffs = append(diffs, fmt.Sprintf("MachineSet %s: diff: %s", ms.Name, strings.Join(notUpToDateResult.LogMessages, ", ")))
 		}
 	}
 
-	if len(matchingMachineSets) == 0 {
-		return nil, fmt.Sprintf("couldn't find MachineSet matching MachineDeployment spec template: %s", strings.Join(diffs, "; ")), nil
+	if len(newMSCandidates) == 0 {
+		return nil, oldMSs, oldMSNotUpToDateResults, fmt.Sprintf("couldn't find MachineSet matching MachineDeployment spec template: %s", strings.Join(diffs, "; "))
```

Comment:
Probably doesn't matter for this PR (but later). I think `EligibleForInPlaceUpdate` is not necessarily set correctly here

(if I'm right, please ensure we have test coverage for this case)